### PR TITLE
Move smoltcp-related code to a separate library

### PR DESCRIPTION
**Pull request by lrh2000**
As discussed in the weekly meeting, for implementing https://github.com/asterinas/asterinas/issues/1267, we want to have a separate library that interacts with the smoltcp internals. This PR is a first step in that direction.

This PR aims to move all smoltcp related code in the Asterinas kernel, mostly the code previously located in `kernel/src/net/iface`, into a separate library. This separate library is tentatively named _bigtcp_.

### The API design

I'll briefly describe the APIs below:

#### The `Iface` trait

The `Iface` trait in bigtcp abstracts network interfaces of different mediums (e.g., IP interface or Ethernet interface). This trait is not intended to be implemented by users of bigtcp. Instead, it can only be implemented inside bigtcp since it is a subtrait of a sealed trait `internal::IfaceInternal`:
```rust
pub trait Iface<E>: internal::IfaceInternal<E> + Send + Sync {
    // ...
}
```

#### The `IpIface` and `EtherIface` types

Users of bigtcp should use `IpIface::new` or `EtherIface::new` (the choice depends on the medium of the network interface) to create instances of `Arc<dyn Iface>`.
 - They accept the device through the `WithDevice` trait, which abstracts the lock and reference count.
 - In other words, `WithDevice` can be implemented for something like `Mutex<Loopback>` and can also be implemented for `Arc<SpinLock<dyn AnyNetworkDevice>>` (a type used by virtio-net).
```rust
// ===== iface/phy/ip.rs

pub struct IpIface<D: WithDevice, E> { /* ... */ }

impl<D: WithDevice, E> IpIface<D, E> {
    pub fn new(driver: D, ip_cidr: IpCidr, ext: E) -> Arc<Self> {
        // ...
    }
}

// ===== iface/phy/ether.rs

pub struct EtherIface<D: WithDevice, E> { /* ... */}

impl<D: WithDevice, E> EtherIface<D, E> {
    pub fn new(driver: D, ether_addr: EthernetAddress, ext: E) -> Arc<Self> {
        // ...
    }
}

// ===== device.rs:

pub trait WithDevice: Send + Sync {
    type Device: Device + ?Sized;

    /// Calls the closure with a mutable reference of [`Device`].
    fn with<F, R>(&self, f: F) -> R
    where
        F: FnOnce(&mut Self::Device) -> R;
}
```

Note: Currently, the main difference between `EtherIface` and `IpIface` is that `EtherIface` uses DHCP to get its IP address. This is actually misleading. In the future, the DHCP implementation should be moved to userspace, and the difference should be that `EtherIface` needs to handle Ethernet protocols (like maintaining the ARP cache) and `IpIface` doesn't (of course).

#### The `E` generic parameter

Sometimes it is desirable to associate more information with a network interface.
 - For example, in Asterinas, each interface has a `WaitQueue` on which the background polling thread will sleep.
 - This information should not be placed in bigtcp because (i) it is not a property of an interface (ii) we have future possibilities to extend the polling mechanism by using timers or soft IRQs, which are only accessible in the Asterinas kernel.

To allow this possibility, the `Iface` trait also accepts a generic parameter that can be specified by users of bigtcp:
```rust
// ===== In bigtcp:

impl<E> dyn Iface<E> {
    /// Gets the extension of the iface.
    pub fn ext(&self) -> &E {
        self.common().ext()
    }
}

// ===== In Asterinas kernel:

pub type Iface = dyn aster_bigtcp::iface::Iface<ext::IfaceExt>;

pub struct IfaceExt {
    name: String,
    next_poll_at_ms: AtomicU64,
    polling_wait_queue: WaitQueue,
}

impl IfaceEx for Iface {
    fn name(&self) -> &str {
        &self.ext().name
    }

    fn poll(&self) {
        self.raw_poll(&|next_poll| self.ext().schedule_next_poll(next_poll));
    }
}
```

### Limitations

Although the goal of bigtcp is to move all code related to smoltcp internals out of the Asterinas kernel, this goal is not fully achieved.

For example, we're still calling smoltcp's `send_slice`/`recv_slice` directly in the Asterinas kernel, and also checking smoltcp's error reason directly. I've thought about providing a wrapper method in bigtcp to hide all the details related to smoltcp from the Asterinas kernel, but finally found that this provides little benefit.
```rust
        let result = self
            .bound_socket
            .raw_with(|socket: &mut RawTcpSocket| socket.recv_slice(buf));

	match result {
	    /* .. */
            Err(SendError::InvalidState) => /* .. */,
	}
```

For now, and for these simple cases, I still prefer to just re-export necessary types (like `SendError`) from smoltcp in bigtcp, so that Asterinas can still access them (note that smoltcp is no longer a direct dependency of the Asterinas kernel, but an indirect dependency via bigtcp).

### Unresolved questions

 - Should bigtcp be a library (`kernel/libs`) or a component (`kernel/comp`)? It seems to be a library since it does not maintain global states.
 - Do we really want to use the name "bigtcp", or should we try to find a better name other than "bigtcp"?
 - Are there any problems with the API design, or is there room for improvement?

### CODE REVIEW

**Comment by StevenJiang1110** [comment]
> Should bigtcp be a library (kernel/libs) or a component (kernel/comp)? It seems to be a library since it does not maintain global states.

I agree that `bigtcp` should be a library. The component subsystem is responsible for initializing each component in a specific order. Since bigtcp does not manage any global states, it doesn't require its own crate-level initialization method. Therefore, maintaining it as a library is sufficient.

> Do we really want to use the name "bigtcp", or should we try to find a better name other than "bigtcp"?

Is the word 'big'  intended because 'smol' from 'smoltcp' sounds similar to 'small'? The name is acceptable for me. The other option may be `smolnet`, as we are implementing the network module based on 'smoltcp.' However, this name is stolen from [Redleaf](https://github.com/mars-research/redleaf).

File: kernel/src/net/iface/init.rs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    use aster_bigtcp::{iface::EtherIface, wire::EthernetAddress};
+    use aster_network::AnyNetworkDevice;
+    use aster_virtio::device::network::DEVICE_NAME;
+
+    let virtio_net = aster_network::get_device(DEVICE_NAME).unwrap();
+
+    let ether_addr = virtio_net.disable_irq().lock().mac_addr().0;
+
+    struct Wrapper(Arc<SpinLock<dyn AnyNetworkDevice, PreemptDisabled>>);
+
```
Comment (StevenJiang1110) [inline kernel/src/net/iface/init.rs:45]: Why not just use `Arc<SpinLock<dyn AnyNetworkDevice, LocalIrqDisabled>>)`? It seems you will always disable irq when visiting this device?

**Review by StevenJiang1110 — COMMENTED** [review]
The API changes in this PR looks fine to me!

File: kernel/libs/aster-bigtcp/src/socket/event.rs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// SPDX-License-Identifier: MPL-2.0
+
+/// A callback that will be invoked whenever events occur on the socket.
+pub trait SocketCallback: Send + Sync {
```
[the remark this answers is not in the document]
Comment (lrh2000) [reply kernel/libs/aster-bigtcp/src/socket/event.rs]: I have no preference for the name of this type. So I changed it accordingly.

File: kernel/src/net/iface/init.rs (hunk 1 above)
Comment (lrh2000) [reply kernel/src/net/iface/init.rs:45]: This is for historical reasons. I've added a new commit to take advantage of our new `SpinLock` APIs.

**Review by StevenJiang1110 — COMMENTED** [review]
LGTM!

**Review by StevenJiang1110 — APPROVED** [review]

**Pull request merged**